File: backend/src/main/java/br/com/hematsu/stock_flow/entities/Product.java
package br.com.hematsu.stock_flow.entities;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;

@Entity(name = "tb_products")
public class Product implements Serializable {
    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long productId;
    private String name;
    private String description;
    private String code;
    private Double costPrice;
    private Double salePrice;
    private Integer stockQuantity;

    @ManyToMany
    @JoinTable(name = "tb_product_categories", joinColumns = @JoinColumn(name = "product_id"), inverseJoinColumns = @JoinColumn(name = "category_id"))
    private Set<Category> categories = new HashSet<>();

    public Product() {
    }

    public Product(String name, String description, String code, Double costPrice, Double salePrice,
            Integer stockQuantity) {
        this.name = name;
        this.description = description;
        this.code = code;
        this.costPrice = costPrice;
        this.salePrice = salePrice;
        this.stockQuantity = stockQuantity;
    }

    public Long getProductId() {
        return productId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.code = code;
    }

    public Double getCostPrice() {
        return costPrice;
    }

    public void setCostPrice(Double costPrice) {
        this.costPrice = costPrice;
    }

    public Double getSalePrice() {
        return salePrice;
    }

    public void setSalePrice(Double salePrice) {
        this.salePrice = salePrice;
    }

    public Integer getStockQuantity() {
        return stockQuantity;
    }

    public void setStockQuantity(Integer stockQuantity) {
        this.stockQuantity = stockQuantity;
    }

    public Set<Category> getCategories() {
        return categories;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((productId == null) ? 0 : productId.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Product other = (Product) obj;
        if (productId == null) {
            if (other.productId != null)
                return false;
        } else if (!productId.equals(other.productId))
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "Product [productId=" + productId + ", name=" + name + ", description=" + description + ", code=" + code
                + ", costPrice=" + costPrice + ", salePrice=" + salePrice + ", stockQuantity=" + stockQuantity
                + ", categories=" + categories + "]";
    }
}

File: frontend/lib/models/product.dart
import 'package:frontend/models/category.dart';

class Product {
  Product({
    required this.name,
    required this.code,
    required this.costPrice,
    required this.salePrice,
    required this.stockQuantity,
    required this.categories,
    this.productId,
    this.description,
  });

  final int? productId;
  final String name;
  final String? description;
  final String code;
  final double costPrice;
  final double salePrice;
  final int stockQuantity;
  final List<Category> categories;

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      productId: json['productId'],
      name: json['name'],
      description: json['description'],
      code: json['code'],
      costPrice: (json['costPrice'] as num).toDouble(),
      salePrice: (json['salePrice'] as num).toDouble(),
      stockQuantity: json['stockQuantity'],
      categories: (json['categories'] as List<dynamic>)
          .map((cat) => Category.fromJson(cat))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
        'productId': productId,
        'name': name,
        'description': description,
        'code': code,
        'costPrice': costPrice,
        'salePrice': salePrice,
        'stockQuantity': stockQuantity,
        'categories': categories.map((cat) => cat.toJson()).toList(),
      };
}
